docs/proposals/20230407-optional-cluster-infra-resource.md

```diff
@@ -0,0 +1,207 @@
+- _Kubernetes Cluster Infrastructure_
+  - When we refer to _Kubernetes Cluster Infrastructure_ we aim to distinguish required environmental infrastructure (e.g., cloud virtual networks) in which a Kubernetes cluster resides as a "set of child resources" from the Kubernetes cluster resources themselves (e.g., virtual machines that underlie nodes, managed by Cluster API). Sometimes this is referred to as "BYO Infrastructure"; essentially, we are talking about **infrastructure that supports a Kubernetes cluster, but is not actively managed by Cluster API**. As we will see, this boundary is different when discussing Managed Kubernetes: more infrastructure resources are not managed by Cluster API when running Managed Kubernetes.
+- e.g.
+  - This just means "For example:"!
+
+## Summary
+
+We propose to make provider `<Infra>Cluster` resources optional in order to better represent Managed Kubernetes scenarios where all _Kubernetes Cluster Infrastructure_ is managed by the service provider, and not by Cluster API.
+
+## Motivation
+
+The implementation of Managed Kubernetes scenarios by Cluster API providers occurred after the architectural design of Cluster API, and thus that design process did not consider these Managed Kubernetes scenarios as a user story. In practice, Cluster API's specification has allowed Managed Kubernetes solutions to emerge that aid running fleets of clusters at scale, with CAPA's `AWSManagedCluster` and `AzureManagedCluster` being notable examples. However, because these Managed Kubernetes solutions arrived after the Cluster API contract was defined, providers have not settled on a consistent rendering of how a "Service-Managed Kubernetes" specification fits into a "Cluster API-Managed Kubernetes" surface area.
+
+One particular part of the existing Cluster API surface area that is inconsistent with most Managed Kubernetes user experiences is the accounting of the [Kubernetes API server](https://kubernetes.io/docs/concepts/overview/components/#kube-apiserver). In the canonical "self-managed" user story that Cluster API addresses, it is the provider implementation of Cluster API (e.g., CAPA) that is responsible for scaffolding the necessary _Kubernetes Cluster Infrastructure_ that is required in order to create the Kubernetes API server (e.g., a Load Balancer and a public IP address). This provider responsibility is declared in the `<Infra>Cluster` resource, and carried out via its controllers; and then finally this reconciliation is synchronized with the parent `Cluster` Cluster API resource.
+
+Because there exist Managed Kubernetes scenarios that handle all _Kubernetes Cluster Infrastructure_ responsibilities themselves, Cluster API's requirement of a `<Infra>Cluster` resource leads to weird implementation decisions, because in these scenarios there is no actual work for a Cluster API provider to do to scaffold _Kubernetes Cluster Infrastructure_.
+
+### Goals
```

Comment:
cc @jackfrancis @richardcase @fabriziopandini ^^

Reply:
I'm not sure about the 2nd option (where we have both an InfraCluster + ControlPlane resource, but only the ControlPlane resource has the controlPlaneEndpoint). But perhaps a scenario like this is what @richardcase has referred to as a way to avoid the "dance" between the two resources for the common endpoint property?

My understanding based on @fabriziopandini's idea is that the key change we're proposing here is the 3rd thing above, where we simply add support for a scenario that omits the InfraCluster (in this scenario the controlPlaneEndpoint is necessarily only exposed in the ControlPlane resource).

Not sure about the ClusterClass-missing scenario (the 4th one...)

Reply:
I think for the purpose of the doc we can ignore the 4th one. It works already and this won't change with this proposal.

Reply:
My understanding is that we need also the 2nd option, which was referred to as "managed Kubernetes with extra infra" in the summary table at the end of the doc.

This option originates from some of the discussion threads like e.g https://docs.google.com/document/d/1rqzZfsO6k_RmOHUxx47cALSr_6SeTG89e9C44-oHHdQ/edit?disco=AAAAsvkZ8xA, where you can read comments like
"I can't speak for Azure but for AWS and GCP there is always extra infra that needs to be created (or used in the BYO case) and so an infra cluster would be needed in this case."

Reply:
@fabriziopandini have we discussed how to actually implement that 2nd option (InfraCluster exists but InfraCluster doesn't have controlPlaneEndpoint)?

the option 3 implementation is easy (we allow no InfraCluster resource at all, so obviously it doesn't have the controlPlaneEndpoint)

How do we make the controlPlaneEndpoint "dance" between InfraCluster and ControlPlane flexible, so that sometimes there is not dance, and it's simply derived from/attached to just the ControlPlane?

Reply:
I think like this: https://github.com/kubernetes-sigs/cluster-api/pull/8500#discussion_r1182785727

There is not much of a "dance"in core CAPI. It's more that InfraCluster and ControlPlane controllers have to be implemented depending on the variant.

Reply:
I think both the InfraCluster and the ControlPlane provider have to be aware of in which variants their CRDs are used / can be used.


An example:
Let's assume today you just use the normal AzureCluster with KubeadmControlPlane. AzureCluster provides the controlPlaneEndpoint, core CAPI propagates it up to Cluster.spec.controlPlaneEndpoint and then KubeadmControlPlane reads it from the Cluster.

In other words. AzureCluster fulfills the normal contract and provides the controlPlaneEndpoint. KubeadmControlPlane is a control plane which does not (and can not) create a control plane endpoint, so it will always just read it from there.

It's simply impossible to combine the KubeadmControlPlane with an InfraCluster which doesn't provide a control plane endpoint, because KubeadmControlPlane is not capable of creating a control plane endpoint.

So in my opinion every CRD should have a fixed behavior. Either it is able to provide a control plane endpoint or it isn't (e.g. KubeadmControlPlane is not, AzureManagedCluster is).

Reply:
Let's assume in AWS we want/need an additional InfraCluster and in Azure we don't. This leads to the following valid combinations:

| InfraCluster      | ControlPlane             |
|-------------------|--------------------------|
| AWSCluster        | KubeadmControlPlane      |
| AWSManagedCluster | AWSManagedControlPlane   |
| AzureCluster      | KubeadmControlPlane      |
| -                 | AzureManagedControlPlane |

Reply:
That makes @sbueringer :+1: 

Relating back to the original managed proposal, the changes we are discussing would make the following changes to the original proposal:

- **Option 2** (Two kinds with a ControlPlane and a pass-through InfraCluster) is no longer a recommended fallback option as the pass through "infra cluster" is no longer needed as the control plane endpoint can be reported back via the control plane (i.e. AzureManagedControlPlane and AWSManagedControlPlane).
-  **Option 3** (Two kinds with a Managed Control Plane and Managed Infra Cluster with Better Separation of Responsibilities) the implementation will be easier as the control plane can be reported back via the controlplane (i.e. GCPManagedControlPlane) and not via the infra cluster (i.e. GCPManagedCluster). This stops what i call the "controller dance" where the infra cluster controller watches the control plane/infra clusters and the control plane controller also watches the infra cluster/controlplane.
- A new option (option 5 if we are additive to the proposal) for "Vanilla Managed" (i.e. Control Plane with No infra cluster) is needed that replaces option 2.
- the recommendations are slightly reworded to:
  - We should still try to stay true to the original separation of concerns between infra & control plane 
  - If you have additional infrastructure then Option 3 is the recommendation (as it currently is)
  - If you have no additional infrastructure then the new "Vanilla Managed" (i.e. Control Plane with No infra cluster)  is the recommendation (this replaces the fallback recommendation of option 2)
- We should also specifically define the use cases that @fabriziopandini added to the [doc](https://docs.google.com/document/d/1rqzZfsO6k_RmOHUxx47cALSr_6SeTG89e9C44-oHHdQ/edit#heading=h.j8gx58ylxx0) we used for discussions, as this really helps to frame what we are talking about 

Based on that, I see the following as the current state and the changes required for providers, short term and long term:

|Provider   | Short Term Changes  | Long term changes  |
| ------------- | ------------------------------- | ----------------------------  |
| CAPG       | Report endpoint back via cp & remove logic from infra controller that does this |  N/A                 |
| CAPOCI   | Report endpoint back via cp & remove logic from infra controller that does this .  | N/A                  |
| CAPA       | Report endpoint back via cp  & remove logic from infra controller. Align with new option. | Split additional infra provisioning from control plane to infra cluster. To align with option 3 |
| CAPZ      | Report endpoint back via cp  & remove logic from infra controller. Align with new option.  |     |

Specifically for CAPA, the easiest option would be for us to stay with the "Vanilla Managed" (i.e. Control Plane with No infra cluster) option and not move towards option 3. But we have a bunch of infra provisioning like networking, bastion host creation that doesn't fit into the "control plane" and should really be in "infra cluster".

Reply:
@sbueringer thank you for the clarity 💯 

@richardcase this summary makes sense to me as well, it'll be in this proposal soon :)

My only addition would be that for folks wishing to evolve from the existing option 2 (pass-through <Infra>ManagedCluster) to option 3 (decomposition of Managed Kubernetes configuration and state across <Infra>ManagedCluster and <Infra>ManagedControlPlane there is actually a back-compat story (for example, providers could deprecate certain properties in <Infra>ManagedControlPlane, though still allow them for back-compat, while introducing equivalent properties in <Infra>ManagedCluster, and only allowing one or the other, say, a wehook that validates that equivalent properties aren't in both places). I think the back-compat story for option 2 --> new, vanilla cluster w/ only <Infra>ManagedControlPlane is pretty trivial, as the <Infra>ManagedCluster has only been a superficial requirement if you're using option 2.